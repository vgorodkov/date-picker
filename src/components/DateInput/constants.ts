export const cursorRanges = [
  { start: 0, end: 2 },
  { start: 3, end: 5 },
  { start: 6, end: 10 },
];

export const BACKSPACE_KEY = 'Backspace';

export const DIGITS_REGEXP = /^[0-9]/g;
