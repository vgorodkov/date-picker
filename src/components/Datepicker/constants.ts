export const DEFAULT_DATELIMIT = {
  min: { day: 1, month: 1, year: 2020 },
  max: { day: 1, month: 1, year: 2030 },
};
