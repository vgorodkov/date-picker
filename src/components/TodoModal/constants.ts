export const TODO_LENGTH_LIMIT = 30;
