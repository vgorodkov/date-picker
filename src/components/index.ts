export { Datepicker } from './Datepicker';
export { Rangepicker } from './Rangepicker';
