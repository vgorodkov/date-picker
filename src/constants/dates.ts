import { getTimestampByDate } from '@/utils/getTimestampByDate';

export const CURRENT_DAY = new Date().getDate();

export const CURRENT_YEAR = new Date().getFullYear();

export const CURRENT_MONTH = new Date().getMonth() + 1;

export const WEEK_DAYS = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];

export const MONTHS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'Jule',
  'August',
  'September',
  'October',
  'November',
  'December',
];

export const WEEKS_AMOUNT = 6;

export const DEFAULT_DAY = {
  day: CURRENT_DAY,
  month: CURRENT_MONTH,
  year: CURRENT_YEAR,
};

export const DEFAULT_DAY_WITH_TIMESTAMP = {
  ...DEFAULT_DAY,
  timestamp: getTimestampByDate(DEFAULT_DAY),
};
