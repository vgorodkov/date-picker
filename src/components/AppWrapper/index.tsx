import { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import { ErrorBoundary } from '@/components/ErrorBoundary';
import { theme } from '@/constants/theme';
import { GlobalStyle } from '@/styles/global';

export const AppWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <>
      <GlobalStyle />
      <ErrorBoundary>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </ErrorBoundary>
    </>
  );
};
