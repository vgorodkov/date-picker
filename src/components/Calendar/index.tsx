import { CalendarContent } from '@/components/Calendar/CalendarContent';
import { CalendarHeader } from '@/components/Calendar/CalendarHeader';
import { MONTHS } from '@/constants/dates';
import { useCalendarControl } from '@/hooks/useCalendarControl';
import { getMonthDates } from '@/utils/getMonthDates';
import { geetWeekDatesByWeekIndex } from '@/utils/getWeekDates';

import { CalendarContainer } from './styled';
import { CalendarProps } from './types';

export const Calendar = ({
  selectedDate = null,
  onDateClick,
  firstDayOfWeek = 'Mo',
  range = {
    start: null,
    end: null,
  },
  showHolidays = false,
  holidayTimestamps = [],
  calendarVariant = 'month',
  withTodo = false,
}: CalendarProps) => {
  const { calendarMonth, calendarYear, weekIndex, selectNextPeriod, selectPrevPeriod } =
    useCalendarControl(selectedDate, calendarVariant);

  const calendarMonthName =
    calendarMonth === 0 ? MONTHS[calendarMonth] : MONTHS[+calendarMonth - 1];
  const isWeekMode = calendarVariant === 'week';
  const monthDates = getMonthDates(calendarYear, calendarMonth, firstDayOfWeek);
  const dates = isWeekMode ? geetWeekDatesByWeekIndex(monthDates, weekIndex) : monthDates;

  return (
    <CalendarContainer>
      <CalendarHeader
        selectNextPeriod={selectNextPeriod}
        selectPrevPeriod={selectPrevPeriod}
        calendarMonthName={calendarMonthName}
        calendarYear={calendarYear}
        firstDayOfWeek={firstDayOfWeek}
      />
      <CalendarContent
        dates={dates}
        calendarMonth={calendarMonth}
        onDateClick={onDateClick}
        selectedDate={selectedDate}
        range={range}
        showHolidays={showHolidays}
        holidayTimestamps={holidayTimestamps}
        withTodo={withTodo}
      />
    </CalendarContainer>
  );
};
