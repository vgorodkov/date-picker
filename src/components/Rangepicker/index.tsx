import { useMemo, useState } from 'react';

import { Calendar } from '@/components/Calendar';
import { DateInput } from '@/components/DateInput';
import { Wrapper } from '@/components/Wrapper';
import { DATE_MASK } from '@/constants/dates';
import { spacing } from '@/constants/spacing';
import { FlexContainer, PickerContainer, RelativeContainer } from '@/styles/containers';
import { DateInputValue, MonthDate } from '@/types/date';
import { PickerProps } from '@/types/picker';
import { RangeVariant } from '@/types/range';
import { getTimestampByDate } from '@/utils/getTimestampByDate';
import { isRangeValid } from '@/utils/isRangeValid';
import { transformDateInputToMonthDate } from '@/utils/transformDateInputToMonthDate';

export const Rangepicker = ({
  firstDayOfWeek,
  showHolidays,
  holidayTimestamps,
  calendarVariant,
}: PickerProps) => {
  const [isCalendarOpen, setIsCalendarOpen] = useState(false);
  const [rangeVariant, setRangeVariant] = useState<RangeVariant | null>(null);
  const [startRange, setStartRange] = useState<DateInputValue>(DATE_MASK);
  const [endRange, setEndRange] = useState<DateInputValue>(DATE_MASK);

  const rangeTimestamps = useMemo(() => {
    const start = getTimestampByDate(transformDateInputToMonthDate(startRange));
    const end = getTimestampByDate(transformDateInputToMonthDate(endRange));

    return {
      start,
      end,
    };
  }, [startRange, endRange]);

  const isRangeTimestampsValid = useMemo(() => {
    return isRangeValid(rangeTimestamps);
  }, [rangeTimestamps]);

  const onCalendarDateClick = (selectedDate: MonthDate) => {
    if (rangeVariant === RangeVariant.START) {
      if (selectedDate.timestamp > rangeTimestamps.end) {
        return;
      }
      setStartRange(selectedDate);
    } else if (rangeVariant === RangeVariant.END) {
      if (selectedDate.timestamp < rangeTimestamps.start) {
        return;
      }
      setEndRange(selectedDate);
    }
  };

  const onStartRangeInputClick = () => {
    setRangeVariant(RangeVariant.START);
    setIsCalendarOpen(true);
  };

  const onEndRangeInputClick = () => {
    setRangeVariant(RangeVariant.END);
    setIsCalendarOpen(true);
  };

  return (
    <Wrapper>
      <PickerContainer>
        <FlexContainer $flexFlow="column nowrap" $gap={spacing.s}>
          <DateInput
            isDateValid={isRangeTimestampsValid}
            isSelected={rangeVariant === RangeVariant.START}
            label="From"
            value={startRange}
            setValue={setStartRange}
            onClick={onStartRangeInputClick}
          />
          <DateInput
            isDateValid={isRangeTimestampsValid}
            isSelected={rangeVariant === RangeVariant.END}
            label="To"
            value={endRange}
            setValue={setEndRange}
            onClick={onEndRangeInputClick}
          />
        </FlexContainer>
        <RelativeContainer>
          {isCalendarOpen && (
            <Calendar
              withTodo={false}
              range={isRangeTimestampsValid ? rangeTimestamps : undefined}
              calendarVariant={calendarVariant}
              onDateClick={onCalendarDateClick}
              firstDayOfWeek={firstDayOfWeek}
              showHolidays={showHolidays}
              holidayTimestamps={holidayTimestamps}
            />
          )}
        </RelativeContainer>
      </PickerContainer>
    </Wrapper>
  );
};
