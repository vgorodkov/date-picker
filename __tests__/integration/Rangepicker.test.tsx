import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import { Rangepicker } from '@/components/Rangepicker';
import { CURRENT_MONTH } from '@/constants/dates';

describe('Rangepicker Component', () => {
  it('renders DateInput components correctly', async () => {
    render(<Rangepicker calendarVariant="month" />);

    expect(await screen.findByText(/From/i)).toBeInTheDocument();
    expect(await screen.findByText(/To/i)).toBeInTheDocument();
  });

  test('opens Calendar component on From DateInput click', async () => {
    const { getByTestId } = render(<Rangepicker calendarVariant="month" />);
    const fromDateInputContainer = getByTestId('From-input-container');
    fireEvent.click(fromDateInputContainer);
    await waitFor(() => {
      const calendar = getByTestId('calendar-component');
      expect(calendar).toBeInTheDocument();
    });
  });

  test('selects range on Calendar date click', async () => {
    const { getByTestId } = render(<Rangepicker calendarVariant="month" />);
    const fromDateInputContainer = getByTestId('From-input-container');
    const toDateInputContainer = getByTestId('To-input-container');
    fireEvent.click(fromDateInputContainer);
    await waitFor(() => {
      const firstDate = getByTestId(`calendar-date-1-${CURRENT_MONTH}`);
      fireEvent.click(firstDate);
    });
    fireEvent.click(toDateInputContainer);
    await waitFor(() => {
      const secondDate = getByTestId(`calendar-date-2-${CURRENT_MONTH}`);
      fireEvent.click(secondDate);
    });
    const selectedFromDate = fromDateInputContainer.querySelector('input')?.value;
    const selectedToDate = toDateInputContainer.querySelector('input')?.value;
    expect(selectedFromDate).toBeTruthy();
    expect(selectedToDate).toBeTruthy();
  });
  test('failts to select invalid range on Calendar date click', async () => {
    const { getByTestId } = render(<Rangepicker calendarVariant="month" />);
    const fromDateInputContainer = getByTestId('From-input-container');
    const toDateInputContainer = getByTestId('To-input-container');
    fireEvent.click(fromDateInputContainer);
    await waitFor(() => {
      const fromDate = getByTestId(`calendar-date-5-${CURRENT_MONTH}`);
      fireEvent.click(fromDate);
    });
    fireEvent.click(toDateInputContainer);
    await waitFor(() => {
      const toDate = getByTestId(`calendar-date-1-${CURRENT_MONTH}`);
      fireEvent.click(toDate);
    });
    const selectedFromDate = fromDateInputContainer.querySelector('input')?.value;
    const selectedToDate = toDateInputContainer.querySelector('input')?.value;
    expect(selectedFromDate).toBeTruthy();
    expect(selectedToDate).toEqual('DD/MM/YYYY');
  });
  it('matches the snapshot', () => {
    const component = render(<Rangepicker calendarVariant="month" />);
    expect(component).toMatchSnapshot();
  });
});
