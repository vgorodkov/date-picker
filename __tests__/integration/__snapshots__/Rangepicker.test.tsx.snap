// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rangepicker Component matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-eDLKkx iIGyiZ"
      >
        <div
          class="sc-guDLey bAKPwX"
          flexflow="column nowrap"
          gap="8px"
        >
          <div
            class="sc-guDLey bAKPwX"
            flexflow="column nowrap"
            gap="8px"
          >
            <div
              class="sc-qZrbh dCuewT"
            >
              From
            </div>
            <div
              class="sc-jsEeTM bOsTEI"
              data-testid="From-input-container"
            >
              <img
                class="sc-jTQCzO iYVWsV"
                src="[object Object]"
              />
              <input
                class="sc-kFCroH fKARsL"
                data-testid="date-input"
                placeholder="Select date"
                readonly=""
                type="text"
                value="DD/MM/YYYY"
              />
              <img
                class="sc-jTQCzO iYVWsV"
                src="[object Object]"
              />
            </div>
          </div>
          <div
            class="sc-guDLey bAKPwX"
            flexflow="column nowrap"
            gap="8px"
          >
            <div
              class="sc-qZrbh dCuewT"
            >
              To
            </div>
            <div
              class="sc-jsEeTM bOsTEI"
              data-testid="To-input-container"
            >
              <img
                class="sc-jTQCzO iYVWsV"
                src="[object Object]"
              />
              <input
                class="sc-kFCroH fKARsL"
                data-testid="date-input"
                placeholder="Select date"
                readonly=""
                type="text"
                value="DD/MM/YYYY"
              />
              <img
                class="sc-jTQCzO iYVWsV"
                src="[object Object]"
              />
            </div>
          </div>
        </div>
        <div
          class="sc-dmyCSP bFbSAt"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-eDLKkx iIGyiZ"
    >
      <div
        class="sc-guDLey bAKPwX"
        flexflow="column nowrap"
        gap="8px"
      >
        <div
          class="sc-guDLey bAKPwX"
          flexflow="column nowrap"
          gap="8px"
        >
          <div
            class="sc-qZrbh dCuewT"
          >
            From
          </div>
          <div
            class="sc-jsEeTM bOsTEI"
            data-testid="From-input-container"
          >
            <img
              class="sc-jTQCzO iYVWsV"
              src="[object Object]"
            />
            <input
              class="sc-kFCroH fKARsL"
              data-testid="date-input"
              placeholder="Select date"
              readonly=""
              type="text"
              value="DD/MM/YYYY"
            />
            <img
              class="sc-jTQCzO iYVWsV"
              src="[object Object]"
            />
          </div>
        </div>
        <div
          class="sc-guDLey bAKPwX"
          flexflow="column nowrap"
          gap="8px"
        >
          <div
            class="sc-qZrbh dCuewT"
          >
            To
          </div>
          <div
            class="sc-jsEeTM bOsTEI"
            data-testid="To-input-container"
          >
            <img
              class="sc-jTQCzO iYVWsV"
              src="[object Object]"
            />
            <input
              class="sc-kFCroH fKARsL"
              data-testid="date-input"
              placeholder="Select date"
              readonly=""
              type="text"
              value="DD/MM/YYYY"
            />
            <img
              class="sc-jTQCzO iYVWsV"
              src="[object Object]"
            />
          </div>
        </div>
      </div>
      <div
        class="sc-dmyCSP bFbSAt"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
